File: asticaVoice/python/asticaVoice_tts_stream_http.py
#!/usr/bin/env python3
"""
HTTP streaming TTS example for astica Voice API (Python)

- Calls HTTPS POST https://voice.astica.ai/api/tts with stream=true
- For expressive GPU voices (e.g. "expressive_sarah"), the server returns
  raw 16‑bit PCM audio (audio/pcm), ideal for low-latency playback.
- For neural/programmatic voices, the server returns WAV (audio/wav).

This script:
  - Streams audio over HTTP as it arrives.
  - Optionally plays it live via sounddevice (low-latency).
  - Saves:
      - WAV audio to:  audio-stream-http.wav
      - Summary JSON to: audio-stream-http.json

Setup:
    pip install -r requirements.txt

Usage:
    python tts_stream_http.py
"""

import json
import sys
import wave
from pathlib import Path

import requests

try:
    import sounddevice as sd
except ImportError:
    sd = None


# ---------------------------------------------------------------------------
# Configuration (edit these)
# ---------------------------------------------------------------------------

ASTICA_API_KEY = "YOUR_API_KEY_HERE"  # obtain free api key -> https://astica.ai/api-keys/
ASTICA_API_ENDPOINT = "https://voice.astica.ai"

VOICE = "expressive_sarah"
# Other examples:
#   "prog_avery"      - OpenAI / programmable voice
#   "neural_jennifer" - Azure / neural voice
# -> view all voices: https://astica.ai/voice/text-to-speech/

TEXT = "Hello from astica! This is an example in Python."

# Stream options
PLAY_AUDIO_LIVE = True  # Attempts low-latency playback while streaming
HTTP_TIMEOUT_SECONDS = 60

# Output files
OUTPUT_WAV_PATH = Path("audio-stream-http.wav")
OUTPUT_JSON_PATH = Path("audio-stream-http.json")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main() -> int:
    if not ASTICA_API_KEY or ASTICA_API_KEY == "YOUR_API_KEY_HERE":
        print("Please set ASTICA_API_KEY at the top of this file.")
        return 1

    if PLAY_AUDIO_LIVE and sd is None:
        print("sounddevice is not installed; live playback disabled.")
        print("Install with: pip install sounddevice")
        play_audio = False
    else:
        play_audio = PLAY_AUDIO_LIVE

    url = f"{ASTICA_API_ENDPOINT}/api/tts"
    payload = {
        "tkn": ASTICA_API_KEY,
        "text": TEXT,
        "voice": VOICE,
        "stream": True,
        "timestamps": False,
    }

    print("Calling astica Voice TTS (HTTP streaming)...")
    print(f"  Endpoint: {url}")
    print(f"  Voice:    {VOICE}")
    print(f"  Text:     {TEXT!r}")
    print()

    try:
        resp = requests.post(url, json=payload, stream=True, timeout=HTTP_TIMEOUT_SECONDS)
    except requests.RequestException as e:
        print(f"[Network/HTTP error] {e}")
        return 1

    if not resp.ok:
        try:
            err_json = resp.json()
        except Exception:
            err_json = None

        msg = f"HTTP {resp.status_code}"
        if err_json and isinstance(err_json, dict):
            server_err = err_json.get("error") or err_json.get("status")
            if server_err:
                msg += f" – server error: {server_err}"
        print(msg)
        return 1

    content_type = (resp.headers.get("Content-Type") or "").lower()
    print(f"Server Content-Type: {content_type}")

    audio_format = None
    sample_rate = None

    # -------------------- Expressive (audio/pcm) ------------------------
    if "audio/pcm" in content_type:
        audio_format = "pcm_s16le"
        sample_rate = 24000  # GPU expressive default; server uses 24 kHz mono

        pcm_data = bytearray()
        stream = None

        if play_audio:
            try:
                stream = sd.RawOutputStream(
                    samplerate=sample_rate,
                    channels=1,
                    dtype="int16",
                    blocksize=0,
                    latency="low",
                )
                stream.start()
            except Exception as e:
                print(f"Could not open audio device for playback: {e}")
                print("Continuing without live playback.")
                stream = None

        try:
            for chunk in resp.iter_content(chunk_size=4096):
                if not chunk:
                    continue
                pcm_data.extend(chunk)
                if stream is not None:
                    stream.write(chunk)
        finally:
            if stream is not None:
                stream.stop()
                stream.close()

        # Save as WAV (wrap raw PCM with header)
        with wave.open(str(OUTPUT_WAV_PATH), "wb") as wf:
            wf.setnchannels(1)
            wf.setsampwidth(2)  # 16-bit
            wf.setframerate(sample_rate)
            wf.writeframes(pcm_data)

    # -------------------- Neural / Programmable (audio/wav) -------------
    elif "audio/wav" in content_type:
        audio_format = "wav"
        header_buf = bytearray()
        header_parsed = False
        stream = None

        # Save raw bytes directly as WAV
        with open(OUTPUT_WAV_PATH, "wb") as f:
            try:
                for chunk in resp.iter_content(chunk_size=4096):
                    if not chunk:
                        continue
                    f.write(chunk)

                    if not play_audio:
                        continue

                    if not header_parsed:
                        header_buf.extend(chunk)
                        if len(header_buf) < 44:
                            continue

                        # Parse WAV header from first 44 bytes
                        header = header_buf[:44]
                        # Sample rate is at offset 24 (4 bytes, little-endian)
                        sample_rate = int.from_bytes(header[24:28], "little")
                        num_channels = int.from_bytes(header[22:24], "little")
                        bits_per_sample = int.from_bytes(header[34:36], "little")

                        if bits_per_sample != 16:
                            print(f"Unexpected bits_per_sample={bits_per_sample}, playback may fail.")
                        if num_channels != 1:
                            print(f"Unexpected channels={num_channels}, playback may fail.")

                        first_pcm = header_buf[44:]

                        try:
                            stream = sd.RawOutputStream(
                                samplerate=sample_rate,
                                channels=num_channels,
                                dtype="int16",
                                blocksize=0,
                                latency="low",
                            )
                            stream.start()
                            if first_pcm:
                                stream.write(first_pcm)
                        except Exception as e:
                            print(f"Could not open audio device for playback: {e}")
                            print("Continuing without live playback.")
                            stream = None

                        header_parsed = True
                    else:
                        if stream is not None:
                            stream.write(chunk)
            finally:
                if stream is not None:
                    stream.stop()
                    stream.close()
    else:
        print("Unexpected Content-Type; not audio/pcm or audio/wav.")
        # Still drain the response to avoid connection issues
        _ = resp.content
        return 1

    # ------------------------------------------------------------------
    # Save simple summary JSON
    # ------------------------------------------------------------------
    summary = {
        "endpoint": url,
        "voice": VOICE,
        "text": TEXT,
        "stream": True,
        "content_type": content_type,
        "audio_format": audio_format,
        "sample_rate": sample_rate,
        "note": (
            "Streaming HTTP does not return billing metadata (cost_units). "
            "Use non-streaming (/api/tts with stream=false) or WebSockets (/ws/api) "
            "if you need per-request billing info."
        ),
    }

    OUTPUT_JSON_PATH.write_text(json.dumps(summary, indent=2), encoding="utf-8")

    print()
    print("HTTP streaming TTS completed.")
    print(f"  Audio format: {audio_format}")
    print(f"  Sample rate:  {sample_rate}")
    print()
    print(f"Saved audio to: {OUTPUT_WAV_PATH.resolve()}")
    print(f"Saved JSON  to: {OUTPUT_JSON_PATH.resolve()}")

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: asticaVoice/node/asticaVoice_tts_stream_http.js
#!/usr/bin/env node
/**
 * HTTP streaming TTS example for astica Voice API (Node.js)
 *
 * - Calls HTTPS POST https://voice.astica.ai/api/tts with stream=true
 * - For expressive GPU voices ("expressive_sarah"), server returns raw PCM (audio/pcm).
 * - For neural/programmatic voices, server returns WAV (audio/wav).
 *
 * This script:
 *   - Streams audio as it arrives and (optionally) plays it live using `speaker`.
 *   - Saves final audio to:  audio-stream-http.wav
 *   - Saves a small JSON summary to: audio-stream-http.json
 *
 * Requirements:
 *   - Node.js 18+ (global fetch)
 *   - Optional: `npm install speaker` for live playback
 *
 * Usage:
 *   node tts_stream_http.js
 */

"use strict";

const fs = require("fs");
const path = require("path");

// Optional dependency for live playback
let Speaker = null;
try {
  // eslint-disable-next-line global-require
  Speaker = require("speaker");
} catch {
  Speaker = null;
  console.log("WARNING: Node cannot play audio. Please 'npm install speaker'");
}


// ---------------------------------------------------------------------------
// Configuration (edit these)
// ---------------------------------------------------------------------------

const ASTICA_API_KEY = "YOUR_API_KEY_HERE"; // obtain free api key -> https://astica.ai/api-keys/
const ASTICA_API_ENDPOINT = "https://voice.astica.ai";

const VOICE = "expressive_sarah";
// Other examples:
//   "prog_avery"      - OpenAI / programmable voice
//   "neural_jennifer" - Azure / neural voice
// -> view all voices: https://astica.ai/voice/text-to-speech/

const TEXT = "Hello from astica! This is an example in Node.";

// Stream options
const PLAY_AUDIO_LIVE = true;   // Try to play audio as it streams (if `npm install speaker` is available)
const HTTP_TIMEOUT_MS = 60_000;

// Output files
const OUTPUT_WAV_PATH = path.resolve("audio-stream-http.wav");
const OUTPUT_JSON_PATH = path.resolve("audio-stream-http.json");


// ---------------------------------------------------------------------------
// Helper: fetch with timeout
// ---------------------------------------------------------------------------

async function fetchWithTimeout(url, options = {}, timeoutMs = 30000) {
  const controller = new AbortController();
  const id = setTimeout(() => controller.abort(), timeoutMs);

  try {
    const resp = await fetch(url, { ...options, signal: controller.signal });
    return resp;
  } finally {
    clearTimeout(id);
  }
}


// ---------------------------------------------------------------------------
// Helper: build WAV header for raw PCM (16‑bit mono)
// ---------------------------------------------------------------------------

function buildWavHeader(pcmByteLength, sampleRate, numChannels = 1, bitsPerSample = 16) {
  const header = Buffer.alloc(44);
  const blockAlign = (numChannels * bitsPerSample) / 8;
  const byteRate = sampleRate * blockAlign;
  const dataLen = pcmByteLength;
  const riffSize = 36 + dataLen;

  header.write("RIFF", 0);
  header.writeUInt32LE(riffSize, 4);
  header.write("WAVE", 8);
  header.write("fmt ", 12);
  header.writeUInt32LE(16, 16); // PCM header size
  header.writeUInt16LE(1, 20);  // PCM format
  header.writeUInt16LE(numChannels, 22);
  header.writeUInt32LE(sampleRate, 24);
  header.writeUInt32LE(byteRate, 28);
  header.writeUInt16LE(blockAlign, 32);
  header.writeUInt16LE(bitsPerSample, 34);
  header.write("data", 36);
  header.writeUInt32LE(dataLen, 40);

  return header;
}


// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main() {
  if (!ASTICA_API_KEY || ASTICA_API_KEY === "YOUR_API_KEY_HERE") {
    console.error("Please set ASTICA_API_KEY at the top of this file.");
    process.exitCode = 1;
    return;
  }

  const canPlayLive = PLAY_AUDIO_LIVE && !!Speaker;

  if (PLAY_AUDIO_LIVE && !Speaker) {
    console.log("`speaker` module not installed; live playback disabled.");
    console.log("Install with: npm install speaker");
  }

  const url = `${ASTICA_API_ENDPOINT}/api/tts`;
  const payload = {
    tkn: ASTICA_API_KEY,
    text: TEXT,
    voice: VOICE,
    stream: true,
    timestamps: false
  };

  console.log("Calling astica Voice TTS (HTTP streaming)...");
  console.log(`  Endpoint: ${url}`);
  console.log(`  Voice:    ${VOICE}`);
  console.log(`  Text:     ${JSON.stringify(TEXT)}`);
  console.log();

  let resp;
  try {
    resp = await fetchWithTimeout(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      },
      body: JSON.stringify(payload)
    }, HTTP_TIMEOUT_MS);
  } catch (e) {
    console.error("[Network/HTTP error]", e.message || e);
    process.exitCode = 1;
    return;
  }

  if (!resp.ok) {
    let bodyText = "";
    let errJson = null;
    try {
      bodyText = await resp.text();
      errJson = JSON.parse(bodyText);
    } catch {
      // ignore
    }

    let msg = `HTTP ${resp.status}`;
    if (errJson && typeof errJson === "object") {
      const serverErr = errJson.error || errJson.status;
      if (serverErr) {
        msg += ` – server error: ${serverErr}`;
      }
    } else if (bodyText) {
      msg += ` – body: ${bodyText}`;
    }
    console.error(msg);
    process.exitCode = 1;
    return;
  }

  const contentType = (resp.headers.get("content-type") || "").toLowerCase();
  console.log(`Server Content-Type: ${contentType}`);

  let audioFormat = null;
  let sampleRate = null;

  if (!resp.body) {
    console.error("No response body received (no streaming).");
    process.exitCode = 1;
    return;
  }

  // Expressive GPU (audio/pcm)
  if (contentType.includes("audio/pcm")) {
    audioFormat = "pcm_s16le";
    sampleRate = 24000; // default GPU expressive

    const pcmChunks = [];
    let speaker = null;

    if (canPlayLive) {
      try {
        speaker = new Speaker({
          channels: 1,
          bitDepth: 16,
          sampleRate
        });
      } catch (e) {
        console.warn("Could not open audio device for playback:", e.message || e);
        console.warn("Continuing without live playback.");
        speaker = null;
      }
    }

    const reader = resp.body.getReader();
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      if (!value || !value.length) continue;

      const chunk = Buffer.from(value);
      pcmChunks.push(chunk);

      if (speaker) {
        speaker.write(chunk);
      }
    }

    if (speaker) {
      speaker.end();
    }

    const pcmBuf = Buffer.concat(pcmChunks);
    const header = buildWavHeader(pcmBuf.length, sampleRate, 1, 16);
    const wavBuf = Buffer.concat([header, pcmBuf]);
    fs.writeFileSync(OUTPUT_WAV_PATH, wavBuf);
  }
  // Neural/OpenAI: audio/wav
  else if (contentType.includes("audio/wav")) {
    audioFormat = "wav";

    const outStream = fs.createWriteStream(OUTPUT_WAV_PATH);
    let headerBuf = Buffer.alloc(0);
    let headerParsed = false;
    let speaker = null;

    const reader = resp.body.getReader();
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      if (!value || !value.length) continue;

      const chunk = Buffer.from(value);
      outStream.write(chunk);

      if (!canPlayLive) {
        continue;
      }

      if (!headerParsed) {
        headerBuf = Buffer.concat([headerBuf, chunk]);
        if (headerBuf.length < 44) {
          continue;
        }

        const header = headerBuf.slice(0, 44);
        sampleRate = header.readUInt32LE(24);  // offset 24
        const numChannels = header.readUInt16LE(22);
        const bitsPerSample = header.readUInt16LE(34);
        const firstPcm = headerBuf.slice(44);

        try {
          speaker = new Speaker({
            channels: numChannels,
            bitDepth: bitsPerSample,
            sampleRate
          });
          if (firstPcm.length) {
            speaker.write(firstPcm);
          }
        } catch (e) {
          console.warn("Could not open audio device for playback:", e.message || e);
          console.warn("Continuing without live playback.");
          speaker = null;
        }

        headerParsed = true;
      } else if (speaker) {
        speaker.write(chunk);
      }
    }

    outStream.end();
    if (speaker) {
      speaker.end();
    }
  } else {
    console.error("Unexpected Content-Type; expected audio/pcm or audio/wav.");
    // Drain body
    await resp.arrayBuffer().catch(() => {});
    process.exitCode = 1;
    return;
  }

  // Summary JSON
  const summary = {
    endpoint: url,
    voice: VOICE,
    text: TEXT,
    stream: true,
    content_type: contentType,
    audio_format: audioFormat,
    sample_rate: sampleRate,
    note:
      "Streaming HTTP does not return billing metadata (cost_units). " +
      "Use non-streaming (/api/tts with stream=false) or WebSockets (/ws/api) " +
      "if you need per-request billing info."
  };

  fs.writeFileSync(OUTPUT_JSON_PATH, JSON.stringify(summary, null, 2), "utf8");

  console.log();
  console.log("HTTP streaming TTS completed.");
  console.log(`  Audio format: ${audioFormat}`);
  console.log(`  Sample rate:  ${sampleRate}`);
  console.log();
  console.log(`Saved audio to: ${OUTPUT_WAV_PATH}`);
  console.log(`Saved JSON  to: ${OUTPUT_JSON_PATH}`);
}


if (require.main === module) {
  main().catch((err) => {
    console.error("Unexpected error:", err);
    process.exitCode = 1;
  });
}
